File: plugin/skills/gioia-methodology/scripts/validate-structure.js
#!/usr/bin/env node
/**
 * validate-structure.js
 * Validates a Gioia data structure JSON file against the schema
 *
 * Usage:
 *   node validate-structure.js --structure-path /path/to/data-structure.json
 */

const fs = require('fs');
const path = require('path');

function parseArgs() {
  const args = process.argv.slice(2);
  const parsed = {};

  for (let i = 0; i < args.length; i += 2) {
    const key = args[i].replace('--', '');
    const value = args[i + 1];
    parsed[key] = value;
  }

  return parsed;
}

function validateStructure(structure) {
  const results = {
    valid: true,
    errors: [],
    warnings: [],
    stats: {
      dimensions: 0,
      themes: 0,
      concepts: 0,
      quotes: 0
    }
  };

  // Check for aggregate_dimensions array
  if (!structure.aggregate_dimensions || !Array.isArray(structure.aggregate_dimensions)) {
    results.errors.push('Missing or invalid aggregate_dimensions array');
    results.valid = false;
    return results;
  }

  const seenIds = new Set();

  // Validate each dimension
  for (const dimension of structure.aggregate_dimensions) {
    results.stats.dimensions++;

    // Check dimension fields
    if (!dimension.id) {
      results.errors.push(`Dimension missing id: ${JSON.stringify(dimension).slice(0, 50)}...`);
      results.valid = false;
    } else {
      if (seenIds.has(dimension.id)) {
        results.errors.push(`Duplicate id: ${dimension.id}`);
        results.valid = false;
      }
      seenIds.add(dimension.id);

      // Check ID format (should start with AD)
      if (!dimension.id.startsWith('AD')) {
        results.warnings.push(`Dimension id "${dimension.id}" should start with "AD" (e.g., AD1)`);
      }
    }

    if (!dimension.name || dimension.name.includes('{')) {
      results.errors.push(`Dimension ${dimension.id || 'unknown'}: missing or placeholder name`);
      results.valid = false;
    }

    if (!dimension.definition) {
      results.warnings.push(`Dimension ${dimension.id}: missing definition`);
    }

    // Validate themes
    if (!dimension.second_order_themes || !Array.isArray(dimension.second_order_themes)) {
      results.errors.push(`Dimension ${dimension.id}: missing second_order_themes array`);
      results.valid = false;
      continue;
    }

    if (dimension.second_order_themes.length === 0) {
      results.warnings.push(`Dimension ${dimension.id}: has no themes`);
    }

    for (const theme of dimension.second_order_themes) {
      results.stats.themes++;

      // Check theme fields
      if (!theme.id) {
        results.errors.push(`Theme missing id in dimension ${dimension.id}`);
        results.valid = false;
      } else {
        if (seenIds.has(theme.id)) {
          results.errors.push(`Duplicate id: ${theme.id}`);
          results.valid = false;
        }
        seenIds.add(theme.id);

        // Check ID format (should include parent dimension)
        if (!theme.id.includes('_T')) {
          results.warnings.push(`Theme id "${theme.id}" should follow format "AD1_T1"`);
        }
      }

      if (!theme.name || theme.name.includes('{')) {
        results.errors.push(`Theme ${theme.id || 'unknown'}: missing or placeholder name`);
        results.valid = false;
      }

      if (!theme.researcher_interpretation) {
        results.warnings.push(`Theme ${theme.id}: missing researcher_interpretation (important for audit trail)`);
      }

      // Validate concepts
      if (!theme.first_order_concepts || !Array.isArray(theme.first_order_concepts)) {
        results.errors.push(`Theme ${theme.id}: missing first_order_concepts array`);
        results.valid = false;
        continue;
      }

      if (theme.first_order_concepts.length === 0) {
        results.warnings.push(`Theme ${theme.id}: has no concepts`);
      }

      for (const concept of theme.first_order_concepts) {
        results.stats.concepts++;

        // Check concept fields
        if (!concept.id) {
          results.errors.push(`Concept missing id in theme ${theme.id}`);
          results.valid = false;
        } else {
          if (seenIds.has(concept.id)) {
            results.errors.push(`Duplicate id: ${concept.id}`);
            results.valid = false;
          }
          seenIds.add(concept.id);

          // Check ID format
          if (!concept.id.includes('_C')) {
            results.warnings.push(`Concept id "${concept.id}" should follow format "AD1_T1_C1"`);
          }
        }

        if (!concept.name || concept.name.includes('{')) {
          results.errors.push(`Concept ${concept.id || 'unknown'}: missing or placeholder name`);
          results.valid = false;
        }

        if (!concept.informant_terms || concept.informant_terms.length === 0) {
          results.warnings.push(`Concept ${concept.id}: missing informant_terms (important for 1st-order grounding)`);
        }

        // Count quotes
        if (concept.example_quotes && Array.isArray(concept.example_quotes)) {
          for (const quote of concept.example_quotes) {
            results.stats.quotes++;

            if (!quote.document_id) {
              results.warnings.push(`Quote in ${concept.id}: missing document_id`);
            }
            if (!quote.lines) {
              results.warnings.push(`Quote in ${concept.id}: missing line numbers`);
            }
            if (!quote.quote) {
              results.errors.push(`Quote in ${concept.id}: missing quote text`);
              results.valid = false;
            }
          }
        } else {
          results.warnings.push(`Concept ${concept.id}: no example_quotes (should have at least 2-3)`);
        }
      }
    }
  }

  return results;
}

// Main execution
const args = parseArgs();

if (!args['structure-path']) {
  console.error(JSON.stringify({
    success: false,
    error: 'Missing required argument: --structure-path'
  }));
  process.exit(1);
}

const structurePath = args['structure-path'];

// Path traversal protection - resolve to absolute path
const resolvedPath = path.resolve(structurePath);
if (resolvedPath.includes('\0')) {  // Null byte injection protection
  console.error(JSON.stringify({
    success: false,
    error: 'Invalid path - null bytes detected'
  }));
  process.exit(1);
}

if (!fs.existsSync(resolvedPath)) {
  console.error(JSON.stringify({
    success: false,
    error: `File not found: ${structurePath}`
  }));
  process.exit(1);
}

let structure;
try {
  const content = fs.readFileSync(resolvedPath, 'utf8');
  structure = JSON.parse(content);
} catch (error) {
  console.error(JSON.stringify({
    success: false,
    error: `Failed to parse JSON: ${error.message}`
  }));
  process.exit(1);
}

const results = validateStructure(structure);
console.log(JSON.stringify({
  success: results.valid,
  ...results
}, null, 2));

process.exit(results.valid ? 0 : 1);

File: plugin-kimi/.agents/skills/qual-gioia/scripts/validate_structure.py
#!/usr/bin/env python3
"""
validate_structure.py
Validates a Gioia data structure JSON file against the schema

Usage:
  python validate_structure.py --structure-path /path/to/data-structure.json
"""

import argparse
import json
import sys
from pathlib import Path
from typing import Dict, List, Any, Set


def validate_structure(structure: Dict[str, Any]) -> Dict[str, Any]:
    """Validate a Gioia data structure."""
    results = {
        "valid": True,
        "errors": [],
        "warnings": [],
        "stats": {
            "dimensions": 0,
            "themes": 0,
            "concepts": 0,
            "quotes": 0
        }
    }

    # Check for aggregate_dimensions array
    if not structure.get("aggregate_dimensions") or not isinstance(structure["aggregate_dimensions"], list):
        results["errors"].append("Missing or invalid aggregate_dimensions array")
        results["valid"] = False
        return results

    seen_ids: Set[str] = set()

    # Validate each dimension
    for dimension in structure["aggregate_dimensions"]:
        results["stats"]["dimensions"] += 1

        # Check dimension fields
        dim_id = dimension.get("id")
        if not dim_id:
            results["errors"].append(f"Dimension missing id: {str(dimension)[:50]}...")
            results["valid"] = False
        else:
            if dim_id in seen_ids:
                results["errors"].append(f"Duplicate id: {dim_id}")
                results["valid"] = False
            seen_ids.add(dim_id)

            # Check ID format (should start with AD)
            if not dim_id.startswith("AD"):
                results["warnings"].append(f'Dimension id "{dim_id}" should start with "AD" (e.g., AD1)')

        dim_name = dimension.get("name", "")
        if not dim_name or "{" in dim_name:
            results["errors"].append(f"Dimension {dim_id or 'unknown'}: missing or placeholder name")
            results["valid"] = False

        if not dimension.get("definition"):
            results["warnings"].append(f"Dimension {dim_id}: missing definition")

        # Validate themes
        themes = dimension.get("second_order_themes")
        if not themes or not isinstance(themes, list):
            results["errors"].append(f"Dimension {dim_id}: missing second_order_themes array")
            results["valid"] = False
            continue

        if len(themes) == 0:
            results["warnings"].append(f"Dimension {dim_id}: has no themes")

        for theme in themes:
            results["stats"]["themes"] += 1

            # Check theme fields
            theme_id = theme.get("id")
            if not theme_id:
                results["errors"].append(f"Theme missing id in dimension {dim_id}")
                results["valid"] = False
            else:
                if theme_id in seen_ids:
                    results["errors"].append(f"Duplicate id: {theme_id}")
                    results["valid"] = False
                seen_ids.add(theme_id)

                # Check ID format (should include parent dimension)
                if "_T" not in theme_id:
                    results["warnings"].append(f'Theme id "{theme_id}" should follow format "AD1_T1"')

            theme_name = theme.get("name", "")
            if not theme_name or "{" in theme_name:
                results["errors"].append(f"Theme {theme_id or 'unknown'}: missing or placeholder name")
                results["valid"] = False

            if not theme.get("researcher_interpretation"):
                results["warnings"].append(
                    f"Theme {theme_id}: missing researcher_interpretation (important for audit trail)"
                )

            # Validate concepts
            concepts = theme.get("first_order_concepts")
            if not concepts or not isinstance(concepts, list):
                results["errors"].append(f"Theme {theme_id}: missing first_order_concepts array")
                results["valid"] = False
                continue

            if len(concepts) == 0:
                results["warnings"].append(f"Theme {theme_id}: has no concepts")

            for concept in concepts:
                results["stats"]["concepts"] += 1

                # Check concept fields
                concept_id = concept.get("id")
                if not concept_id:
                    results["errors"].append(f"Concept missing id in theme {theme_id}")
                    results["valid"] = False
                else:
                    if concept_id in seen_ids:
                        results["errors"].append(f"Duplicate id: {concept_id}")
                        results["valid"] = False
                    seen_ids.add(concept_id)

                    # Check ID format
                    if "_C" not in concept_id:
                        results["warnings"].append(
                            f'Concept id "{concept_id}" should follow format "AD1_T1_C1"'
                        )

                concept_name = concept.get("name", "")
                if not concept_name or "{" in concept_name:
                    results["errors"].append(f"Concept {concept_id or 'unknown'}: missing or placeholder name")
                    results["valid"] = False

                if not concept.get("informant_terms"):
                    results["warnings"].append(
                        f"Concept {concept_id}: missing informant_terms (important for 1st-order grounding)"
                    )

                # Count quotes
                quotes = concept.get("example_quotes", [])
                if quotes and isinstance(quotes, list):
                    for quote in quotes:
                        results["stats"]["quotes"] += 1

                        if not quote.get("document_id"):
                            results["warnings"].append(f"Quote in {concept_id}: missing document_id")
                        if not quote.get("lines"):
                            results["warnings"].append(f"Quote in {concept_id}: missing line numbers")
                        if not quote.get("quote"):
                            results["errors"].append(f"Quote in {concept_id}: missing quote text")
                            results["valid"] = False
                else:
                    results["warnings"].append(
                        f"Concept {concept_id}: no example_quotes (should have at least 2-3)"
                    )

    return results


def main():
    parser = argparse.ArgumentParser(description="Validate Gioia data structure")
    parser.add_argument("--structure-path", required=True, help="Path to data structure JSON file")
    args = parser.parse_args()

    structure_path = Path(args.structure_path).resolve()

    # Path traversal protection - check for null bytes
    if "\x00" in str(structure_path):
        print(json.dumps({"success": False, "error": "Invalid path - null bytes detected"}))
        sys.exit(1)

    if not structure_path.exists():
        print(json.dumps({"success": False, "error": f"File not found: {args.structure_path}"}))
        sys.exit(1)

    try:
        with open(structure_path, "r", encoding="utf-8") as f:
            structure = json.load(f)
    except json.JSONDecodeError as e:
        print(json.dumps({"success": False, "error": f"Failed to parse JSON: {e}"}))
        sys.exit(1)
    except Exception as e:
        print(json.dumps({"success": False, "error": f"Failed to read file: {e}"}))
        sys.exit(1)

    results = validate_structure(structure)
    output = {"success": results["valid"], **results}
    print(json.dumps(output, indent=2))

    sys.exit(0 if results["valid"] else 1)


if __name__ == "__main__":
    main()
